File: Data Structures & Algorithms/meeting-schedule-ii/submission-1.js
/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

class Solution {
   /**
    * Time complexity: O(nlogn)
    * Auxiliary space complexity: O(n)
    * Tags: intervals, heap
    * @param {Interval[]} intervals
    * @return {number}
    */
   minMeetingRooms(intervals) {
      intervals.sort((a, b) => a.start - b.start);
      const roomHeap = new MinPriorityQueue();

      for (const interval of intervals) {
         if (roomHeap.size() && roomHeap.front() <= interval.start) {
            roomHeap.pop();
         }
         roomHeap.push(interval.end);
      }
      return roomHeap.size()
   };
}

File: Data Structures & Algorithms/buildings-with-an-ocean-view/submission-1.py
class Solution:
    def findBuildings(self, heights: list[int]) -> list[int]:
        """
        Time complexity: O(n)
        Auxiliary space complexity: O(n)
        Tags:
            DS: monotonic increasing stack
            A: iteration
        """
        stack = []

        for idx, height in enumerate(heights):
            while stack and stack[-1][0] <= height:
                stack.pop()

            stack.append((height, idx))

        return [idx for _, idx in stack]